server/broadcast.go

```diff
@@ -462,6 +501,85 @@ func shouldStopSession(err error) bool {
 	return sessionErrRegex.MatchString(err.Error())
 }
 
+func verify(verifier *verification.SegmentVerifier, cxn *rtmpConnection,
+	sess *BroadcastSession, source *stream.HLSSegment,
+	res *net.TranscodeData, URIs []string) error {
+
+	// Cache segment contents if necessary.
+	// If we need to retry transcoding because verification fails,
+	// the the segments' OS location will be overwritten.
+	// Cache the segments so we can restore them in OS if necessary.
+	renditionData := make([][]byte, len(URIs))
+	for i, fname := range URIs {
+		if sess.BroadcasterOS.IsExternal() && drivers.IsOwnExternal(fname) {
+			// If broadcaster is using external storage and segments are there
+			// Then have the verifier use that external storage.
+			continue
+		}
+
+		// Sanity check for absolute URIs to ensure we're actually local.
+		uri, err := url.ParseRequestURI(fname)
+		if err != nil {
+			return err // Implies this is retryable?
+		}
+		memOS, ok := sess.BroadcasterOS.(*drivers.MemorySession)
+		if !uri.IsAbs() && ok {
+			data := memOS.GetData(fname)
+			if data == nil {
+				return errors.New("Missing Local Data")
+			}
+			renditionData[i] = data
+		} else {
+			return errors.New("Expected local storage but did not have it")
+		}
+	}
+
+	params := &verification.Params{
+		ManifestID:   sess.ManifestID,
+		Source:       source,
+		Profiles:     sess.Profiles,
+		Orchestrator: sess.OrchestratorInfo,
+		Results:      res,
+		URIs:         URIs,
+		Renditions:   renditionData,
+	}
+
+	// The return value from the verifier, if any, are the *accepted* params.
+	// The accepted params are not necessarily the same as `params` sent here.
+	// The accepted params may be from an earlier iteration if max retries hit.
+	accepted, err := verifier.Verify(params)
+	if _, retry := err.(verification.Retryable); retry {
+		// If retryable, means tampering was detected from this O
+		// Remove the O from the working set for now
+		// Error falls through towards end if necessary
+		cxn.sessManager.removeSession(sess)
+	}
+	if accepted != nil {
+		// The returned set of results has been accepted by the verifier
+
+		// Check if an earlier verification attempt was the one accepted.
+		// If so, reset the local OS if we're using that since it's been
+		// overwritten with this rendition.
+		if accepted != params && !sess.BroadcasterOS.IsExternal() {
```

Comment:
The renditions wouldn't be overwritten in an external OS for multiple verification attempts?

Reply:
If an external OS is used, then each orchestrator writes to a different path. This avoids issues around overwriting, front-running, etc. The actual URL which gets inserted into the playlist is the one that passes verification.

https://github.com/livepeer/go-livepeer/blob/775a76ba16fc71bf901030f994b413f682ff3fdd/server/broadcast.go#L215-L220